esp-hal/src/gpio/mod.rs

```diff
@@ -883,22 +891,55 @@ impl InterruptConfigurable for Io {
 
 #[ram]
 extern "C" fn user_gpio_interrupt_handler() {
-    USER_INTERRUPT_HANDLER.call();
-
-    default_gpio_interrupt_handler();
+    handle_pin_interrupts(|| USER_INTERRUPT_HANDLER.call());
 }
 
 #[ram]
 extern "C" fn default_gpio_interrupt_handler() {
-    handle_pin_interrupts(on_pin_irq);
+    handle_pin_interrupts(|| ());
 }
 
 #[ram]
-fn on_pin_irq(pin_nr: u8) {
-    // FIXME: async handlers signal completion by disabling the interrupt, but this
-    // conflicts with user handlers.
-    set_int_enable(pin_nr, 0, 0, false);
-    asynch::PIN_WAKERS[pin_nr as usize].wake(); // wake task
+fn handle_pin_interrupts(user_handler: fn()) {
+    let intrs_bank0 = InterruptStatusRegisterAccess::Bank0.interrupt_status_read();
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    let intrs_bank1 = InterruptStatusRegisterAccess::Bank1.interrupt_status_read();
+
+    user_handler();
+
+    // Clear interrupt bits early so that we catch repeat interrupts with larger
+    // probability.
```

Comment:
If so, should the interrupt bits be clear before the atomic op?

Though I doubt the executor would be fast enough to catch the double wake.

Reply:
On the same core, we'll just handle the interrupt before anything else gets a chance to run - on dual-core it probably matters but if you rely on microsecond pulses to be caught correctly, you're going to have a bad time anyway.